crates/macros/src/implement_macro.rs

```diff
@@ -0,0 +1,167 @@
+use gen::{TypeKind, TypeReader};
+use std::collections::*;
+use syn::parse::*;
+use syn::*;
+
+custom_keyword!(extend);
+
+#[derive(Debug, Default)]
+pub struct ImplementMacro {
+    pub extend: Option<(&'static str, &'static str)>,
+    pub overrides: BTreeSet<&'static str>,
+    pub implement: BTreeSet<(&'static str, &'static str)>,
+}
+
+impl ImplementMacro {
+    fn parse_implement(&mut self, reader: &'static TypeReader, cursor: ParseStream) -> Result<()> {
+        self.walk_implement(reader, &cursor.parse()?, &mut String::new())?;
+
+        if !cursor.is_empty() {
+            cursor.parse::<Token![,]>()?;
+        }
+
+        Ok(())
+    }
+
+    fn walk_implement(
+        &mut self,
+        reader: &'static TypeReader,
+        tree: &UseTree,
+        namespace: &mut String,
+    ) -> Result<()> {
+        match tree {
+            UseTree::Path(input) => {
+                if !namespace.is_empty() {
+                    namespace.push('.');
+                }
+
+                namespace.push_str(&input.ident.to_string());
+                self.walk_implement(reader, &*input.tree, namespace)?;
+            }
+            UseTree::Name(input) => {
+                let name = input.ident.to_string();
+
+                if let Some((namespace, name)) = reader.get_type_name(namespace, &name) {
```

Comment:
You can also write this like:

```suggestion
                let (namespace, name) = reader
                    .get_type_name(namespace, &name)
                    .ok_or_else(|| Error::new_spanned(input, "Type not found in metadata"))?;
```

And ditch the else-return case.

Reply:
Going to leave this one for now - its nice but with the extra error info I've since added its a bit cumbersome so I'm torn. 😕